api/src/utils/merge-version-data.ts

```diff
@@ -0,0 +1,186 @@
+import type { Item, SchemaOverview } from '@directus/types';
+import Joi from 'joi';
+import type { Alterations } from '../types/index.js';
+
+const alterationSchema = Joi.object({
+	create: Joi.array().items(Joi.object().unknown()),
+	update: Joi.array().items(Joi.object().unknown()),
+	delete: Joi.array().items(Joi.string(), Joi.number()),
+});
+
+export function mergeVersionsRaw(item: Item, versionData: Partial<Item>[]) {
+	const result = { ...item };
+
+	for (const versionRecord of versionData) {
+		for (const key in versionRecord) {
+			result[key] = versionRecord[key];
+		}
+	}
+
+	return result;
+}
+
+export function mergeVersionsRecursive(
+	item: Item,
+	versionData: Partial<Item>[],
+	collection: string,
+	schema: SchemaOverview,
+): Item {
+	if (versionData.length === 0) return item;
+
+	return recursiveMerging(item, versionData, collection, schema) as Item;
+}
+
+function recursiveMerging(
+	data: Item,
+	versionData: Partial<Item>[],
+	collection: string,
+	schema: SchemaOverview,
+): unknown {
+	const result = { ...data };
+	const relations = getRelations(collection, schema);
+
+	for (const versionRecord of versionData) {
+		for (const key in data) {
+			if (!versionRecord || key in versionRecord === false) {
```

Comment:
Theoretically, I'm afraid this could be possible - at least I think I've seen a `null` entry out in the wild (will see if I'm able to reproduce...).
But if we account for that, we should adjust the type 👍 Also shouldn't it be placed like
```js
	for (const versionRecord of versionData) {
		if (!versionRecord) continue;
		// ...
```
?

Reply:
Agreed, it should be shifted higher!